lib/Sema/CSDiagnostics.cpp

```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
-
-    auto type = getType(immInfo.first);
-    auto bgt = type ? type->getAs<BoundGenericType>() : nullptr;
-
-    if (bgt && bgt->getDecl() == getASTContext().getKeyPathDecl())
-      message += " is a read-only key path";
-    else if (VD->isCaptureList())
-      message += " is an immutable capture";
-    else if (VD->isImplicit())
-      message += " is immutable";
-    else if (VD->isLet())
-      message += " is a 'let' constant";
-    else if (!VD->isSettable(DC))
-      message += " is a get-only property";
-    else if (!VD->isSetterAccessibleFrom(DC))
-      message += " setter is inaccessible";
-    else {
-      message += " is immutable";
-    }
-
-    emitDiagnostic(Loc, DeclDiagnostic, message)
-        .highlight(immInfo.first->getSourceRange());
-
-    // If there is a masked instance variable of the same type, emit a
-    // note to fixit prepend a 'self.'.
-    if (auto typeContext = DC->getInnermostTypeContext()) {
-      UnqualifiedLookup lookup(VD->getFullName(), typeContext,
-                               getASTContext().getLazyResolver());
-      for (auto &result : lookup.Results) {
-        const VarDecl *typeVar = dyn_cast<VarDecl>(result.getValueDecl());
-        if (typeVar && typeVar != VD && typeVar->isSettable(DC) &&
-            typeVar->isSetterAccessibleFrom(DC) &&
-            typeVar->getType()->isEqual(VD->getType())) {
-          // But not in its own accessor.
-          auto AD =
-              dyn_cast_or_null<AccessorDecl>(DC->getInnermostMethodContext());
-          if (!AD || AD->getStorage() != typeVar) {
-            emitDiagnostic(Loc, diag::masked_instance_variable,
-                           typeContext->getSelfTypeInContext())
-                .fixItInsert(Loc, "self.");
+  Optional<OverloadChoice> choice = immInfo.second;
+
+  // Attempt diagnostics based on the overload choice.
+  if (choice.hasValue()) {
+
+    if (!choice->isDecl()) {
+      if (choice->getKind() == OverloadChoiceKind::KeyPathApplication &&
+          !isa<ApplyExpr>(immInfo.first)) {
+        std::string message = "the key path";
+        if (auto *SE = dyn_cast<SubscriptExpr>(immInfo.first)) {
+          if (auto *tupleExpr = dyn_cast<TupleExpr>(SE->getIndex())) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(tupleExpr->getElement(0))) {
+              auto identifier = DRE->getDecl()->getBaseName().getIdentifier();
+              message = "'" + identifier.str().str() + "'";
+            }
           }
         }
+        emitDiagnostic(Loc, DeclDiagnostic, message + " is a read-only key path")
+            .highlight(immInfo.first->getSourceRange());
+        return true;
+      }
+      return false;
+    } 
+
+    // Otherwise, we cannot resolve this because the available setter candidates
+    // are all mutating and the base must be mutating.  If we dug out a
+    // problematic decl, we can produce a nice tailored diagnostic.
+    if (auto *VD = dyn_cast_or_null<VarDecl>(choice->getDecl())) {
```

Comment:
Another way would be:

```c++
if (!choice->isDecl()) return;
auto VD = dyn_cast<VarDecl>(choice->getDecl);
```

Reply:
Sadly I don't think we can do this either.  If the choice isn't a decl, we don't wanna return, we just want to skip over the next section. Af this, we try to diagnose the issue using information about the expression so that's what we're trying to skip to. If we have to add the level of indentation anyway, we might as well combine the cast and check in one line.

Reply:
Whoops, I misread the code, sorry. Seems like my suggestions are useless because we _don't_ want to return early; we have to emit a different diagnostic.

Reply:
Can use `dyn_cast<VarDecl>` here because you have already checked that choice has a declaration.

Reply:
@theblixguy Yeah, that sounds like it wouldn't be worth the effort given that the change to the readability here would only be marginal. Good idea though 🙂

Reply:
Changed that to use dyn_cast, so this should be all set 👍